File: src/518_Coin_Change_2/Main.java

class Solution {
    //Runtime: 2 ms, faster than 99.97% 
    //Memory Usage: 36.4 MB, less than 81.19% 
    public int change(int amount, int[] coins) {
        int types = coins.length;
        // LinkedList<> dp = new LinkedList(amount+1);
        int [] dp = new int[amount+1];// 不记得怎么进行2维数组的初始化，只好进行空间压缩
        
        // 无限背包问题，dp为types行，amount+1列
        // 第i行j列表示：仅包含coins[0]~coins[i]这些面额的硬币时，总价值为j时的组合个数
        // dp[i][j] = dp[i-1][j-coins[i]] + dp[i-1][j];
        // dp[0][0] = 1
        dp[0] = 1;
        
        for (int t = 0; t < types; ++t) {
            for (int v = coins[t]; v <= amount; ++v) {
                // dp[t][v] = dp[t-1][v-coins[t]] + dp[t-1][v];
                dp[v] = dp[v-coins[t]] + dp[v];
            }
        }
        return dp[amount];
        
        // 测试一下[1,2,5] target = 5
        //   1 2 3 4 5
        // 1 0 0 0 0 0 initial dp
        //  \|\|\|\|\|
        // 1 1 1 1 1 1 只使用面额coins[0]时的组合数
        //  \ \|\|\|\| 
        //    \|\|\|\|
        // 1 1 2 2 3 3 只使用面额coins[0]和coins[1]时，4 = 1,1,1,1, 4 = 2,2, 4=1,2,1
        //  \        |
        //    \      |
        //      \    |
        //        \  |
        //          \|
        //           4 使用coins[0]，coins[1] 和coins[2]时的组合数
    }
}

public class Main {
  //public static int main(String argv) { 这里输入是个[]
  public static void main(String [] args) {
    int rtn;
    Solution gua = new Solution();

    int amount;
    int [] coins = {1,2,5};
    rtn = gua.change(amount = 5, coins);
    //System.print.ln("expact value = 4, actual value = " + rtn);
    System.out.println("expact value = 4, actual value = " + rtn);
  }
}

File: src/518_Coin_Change_2/main.py

from typing import List

class Solution:
    # Runtime: 136 ms, faster than 87.27%
    # Memory Usage: 14.3 MB, less than 86.92%
    def change(self, amount: int, coins: List[int]) -> int:
        # 无限背包问题，第i行表示为从0~i的元素，列为amount
        size = len(coins)
        dp = [0 for j in range(amount+1)] # 进行空间压缩
        dp[0] = 1 # 初始化第0列amount=0，coin数凑出amount的combination数为1，空集嘛

        for i in range(1,size+1): # 把coin的面额都遍历一遍
            # 当前在处理coins[0]~coins[i]这些个面额时的组合情况
            coin = coins[i-1] # 注意是找组合数

            # 既然在小于coin时都继承上一级的值，那么稍微改进一点点，直接从coin开始
            for val in range(coin, amount+1):
                dp[val] = dp[val-coin] + dp[val]
        return dp[amount]

    # Runtime: 436 ms, faster than 35.96%
    # Memory Usage: 30 MB, less than 22.79%
    def change1(self, amount: int, coins: List[int]) -> int:
        # 无限背包问题，第i行表示为从0~i的元素，列为amount
        # 还是先2D数组，再进行空间压缩
        size = len(coins)
        dp = [[0 for j in range(amount+1)] for i in range(size+1)]

        # 来进行初始化，第0列amount=0，coin数凑出amount的combination数为1，空集嘛
        for pos in range(size+1):
            dp[pos][0] = 1 # 初始化而已，没毛病

        for i in range(1,size+1): # 把coin的面额都遍历一遍
            # 当前在处理coins[0]~coins[i]这些个面额时的组合情况
            coin = coins[i-1] # 注意是找组合数

            # 既然在小于coin时都继承上一级的值，那么稍微改进一点点，直接从coin开始，不过前提是需要空间压缩
            # for val in range(coin, amount+1):
            #     dp[i][val] = dp[i][val-coin] + dp[i-1][val]
            # 下面代码也是OK的
            for val in range(1,amount+1):
                if val < coin:
                    # 此时肯定是没办法加入coins[i-1]的，继承0到i-1元素的combination集合
                    dp[i][val] = dp[i-1][val]
                else:
                    dp[i][val] = dp[i][val-coin] + dp[i-1][val]

        return dp[size][amount]
        #    0 1 2 3
        # [] 1 0 0 0
        # 2  1 0 1 0

        #    0 1 2 3 4 5
        # [] 1 0 0 0 0 0
        # 1  1 1 1 1 1 1
        # 2  1 1 2 2 3 3
        # 5  1 1 2 3 3 4

if __name__ == "__main__":
    gua = Solution()
    rtn = gua.change(amount = 5, coins = [1,2,5])
    print("expect result = 4, actual result = %d " % rtn )

    rtn = gua.change(amount = 10, coins = [10])
    print("expect result = 1, actual result = %d " % rtn )

    rtn = gua.change(amount = 3, coins = [2])
    print("expect result = 0, actual result = %d " % rtn )

